### PR TITLE
adding environment tracking back in

### PR DESCRIPTION
**Pull request by ehfeng**
Choosing (my guess) is the least adopted feature in the event processing pipeline.

Ted found that the problem was I was passing in empty `args` list for `client.sadd(org_id, *args)`, which silently threw an error. I now `return` before this happens.

Adding a test for this as well, as well as a bunch of `return` for convenience.

#nochanges

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">You need to update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/receivers/features.py [hunk 1]
```diff
@@ -37,24 +38,21 @@
...
-#     feature_slugs = []
+@event_processed.connect(weak=False)
+def record_event_processed(project, group, event, **kwargs):
+    feature_slugs = []
 
 #     # Platform
 #     if group.platform in manager.location_slugs('language'):
 #         feature_slugs.append(group.platform)
 
-#     elif event.data.get(get_interface('csp')):
```
Comment (ehfeng) [inline src/sentry/receivers/features.py]: I'm removing this. It doesn't work. In fact, it never worked. Lines 44-46 always worked for csp, which is why the tests always passed.

**Comment by ehfeng** [comment]
I went with `record` return False when nothing is modified, because we were always doing so when `created` is False.

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes to build requirements</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/models/featureadoption.py [hunk 2]
```diff
@@ -102,16 +103,20 @@ def get_all_cache(self, organization_id):
         return {int(x) for x in set.union(*[p.value for p in result])}
 
     def bulk_set_cache(self, organization_id, *args):
+        if not args:
+            return
```
Comment (mattrobenolt) [inline src/sentry/models/featureadoption.py]: Missed one here.

File: src/sentry/models/featureadoption.py [hunk 3]
```diff
@@ -130,10 +135,13 @@ def bulk_record(self, organization_id, feature_slugs, **kwargs):
             feature_ids = set([manager.get_by_slug(slug).id for slug in feature_slugs])
         except UnknownFeature as e:
             logger.exception(e)
-            return
+            return False
 
         incomplete_feature_ids = feature_ids - self.get_all_cache(organization_id)
 
+        if not incomplete_feature_ids:
+            return
```
Comment (mattrobenolt) [inline src/sentry/models/featureadoption.py]: And here.

**Review by mattrobenolt — APPROVED** [review]

File: src/sentry/models/featureadoption.py (hunk 2 above)
Comment (ehfeng) [inline src/sentry/models/featureadoption.py]: Damn. I ended up accidentally committing to the old ` feature-adoption-receivers` branch and cherry picked the commit, so I missed these. Thanks for catching.

File: src/sentry/models/featureadoption.py (hunk 3 above)
Comment (ehfeng) [inline src/sentry/models/featureadoption.py]: Fixed.

**Pull request merged**